### PR TITLE
[MAINT] update flake8 configuration to cover doc strings

### PR DESCRIPTION
**Pull request by Remi-Gau**
This PR enforces some of the flake8 check on the doc strings.

The default behavior of flake would not report those errors:
http://www.pydocstyle.org/en/6.3.0/error_codes.html

Relates to #2528
Relates to https://github.com/nilearn/nilearn/issues/3507

Changes proposed in this pull request:

- apply flake8 to doc strings
- add flake8 doc string plugin
- update flake8 and pre-commit config
- update flake8 ci

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -67,13 +68,28 @@ dev =
...
+    dist,
+	nilearn/externals/tempita
+--select=D,E,F,W,C90
+docstring-convention=numpy
+max-line-length = 79
+max_complexity = 43
+per-file-ignores =
+    # missing docstring
+	*/tests/*: D103
+	*/tests/__init__.py: D104
```
Comment (tsalo) [inline setup.cfg:85]: There's some odd spacing here.

**Comment by tsalo** [comment]
A number of the current flake8 failures come from testing functions. Do you think it's worth linting those, or should they be skipped?

**Comment by Remi-Gau** [comment]
> A number of the current flake8 failures come from testing functions. Do you think it's worth linting those, or should they be skipped?

I should have fixed them but it may be worth deciding how much we want to enforce doc strings in tests.

I have silenced some already but I won't fight back if we decided to just no enforce any of the doc string things in the test files.

File: setup.cfg (hunk 1 above)
Comment (Remi-Gau) [reply setup.cfg]: what do you mean? the indentation?

Comment (Remi-Gau) [reply setup.cfg]: should be better now

File: setup.cfg [hunk 2]
```diff
@@ -67,13 +68,29 @@ dev =
...
+    .git,
+    __pycache__,
+	env,
+    build,
+    dist,
+	nilearn/externals/tempita
+--select = D,E,F,W,C90
+docstring-convention = numpy
+max-line-length = 79
+max_complexity = 43
```
Comment (Remi-Gau) [inline setup.cfg:81]: I have set this to this value as it is currently the maximum of the most complex function / method.
This sets an upper bound to how complex the code can get before flake8 throws an error.
There is a handful of function that would need refactoring to bring this value back towards a more acceptable value (around 20 ish).

**Comment by Remi-Gau** [comment]
> > A number of the current flake8 failures come from testing functions. Do you think it's worth linting those, or should they be skipped?
> 
> I should have fixed them but it may be worth deciding how much we want to enforce doc strings in tests.
> 
> I have silenced some already but I won't fight back if we decided to just no enforce any of the doc string things in the test files.

OK was testing things on some other parts of the code base and it does make sense to relax some of the doc strings formatting rules in the tests suite for now. If we change our mind we start enforcing those rules and clean things up one thing at a time.

File: nilearn/decoding/decoder.py [hunk 3]
```diff
@@ -875,7 +879,7 @@ def _predict_dummy(self, n_samples):
 
 @fill_doc
 class Decoder(_BaseDecoder):
-    """A wrapper for popular classification strategies in neuroimaging.
+    r"""A wrapper for popular classification strategies in neuroimaging.
```
Comment (ymzayek) [inline nilearn/decoding/decoder.py:882]: Why use raw string here?

**Comment by ymzayek** [comment]
> > > A number of the current flake8 failures come from testing functions. Do you think it's worth linting those, or should they be skipped?
> > 
> > 
> > I should have fixed them but it may be worth deciding how much we want to enforce doc strings in tests.
> > I have silenced some already but I won't fight back if we decided to just no enforce any of the doc string things in the test files.
> 
> OK was testing things on some other parts of the code base and it does make sense to relax some of the doc strings formatting rules in the tests suite for now. If we change our mind we start enforcing those rules and clean things up one thing at a time.

Yes I would say to leave them for now. This could be related to pytest refactoring I wanted to do #3331

File: nilearn/decoding/decoder.py (hunk 3 above)
Comment (Remi-Gau) [reply nilearn/decoding/decoder.py]: was getting a:
```
nilearn/decoding/decoder.py:882:1: D301 Use r""" if any backslashes in a docstring
```

so I obediently complied. 🙈 

Now pushed this that should be a better fix:
https://github.com/nilearn/nilearn/pull/3534/commits/6b253036a97646bfdbab2d87705956de3d4bf929

Comment (ymzayek) [reply nilearn/decoding/decoder.py]: Oh I see haha. Ok better now thanks!

**Comment by Remi-Gau** [comment]
BTW this should catch as an error the wrong underline length that the doc build would only throw as warning.

File: setup.cfg (hunk 1 above)
Comment (tsalo) [reply setup.cfg]: Thanks!

**Review by bthirion — APPROVED** [review]
LGTM.

**Review by ymzayek — APPROVED** [review]
LGTM thanks @Remi-Gau

**Pull request merged**